java/arcs/core/host/EntityHandleManager.kt

```diff
@@ -61,85 +63,68 @@ class EntityHandleManager(
-     */
-    suspend fun <T : Entity> createSingletonHandle(
-        mode: HandleMode,
-        baseName: String,
-        entitySpec: EntitySpec<T>,
+    suspend fun <T : Entity> createHandle(
+        spec: HandleSpec<T>,
         storageKey: StorageKey,
-        ttl: Ttl = Ttl.Infinite,
-        idGenerator: Id.Generator = Id.Generator.newSession()
-    ): Handle =
-        idGenerator.handleName(baseName).let { name ->
-            SingletonHandle(
-                name = idGenerator.handleName(baseName),
-                entitySpec = entitySpec,
-                storageProxy = singletonStoreProxy(storageKey, entitySpec),
-                entityPreparer = EntityPreparer(name, idGenerator, entitySpec.SCHEMA, ttl, time),
-                dereferencerFactory = dereferencerFactory
-            )
-        }.let { handle ->
-            return when (mode) {
-                HandleMode.Read -> object : ReadSingletonHandle<T> by handle {}
-                HandleMode.Write -> object : WriteSingletonHandle<T> by handle {}
-                HandleMode.ReadWrite -> object : ReadWriteSingletonHandle<T> by handle {}
-            }
-        }
-
-    /**
-     * Creates and returns a new collection handle for a set of [Entity]s.
-     *
-     * @property mode indicates whether the handle is allowed to read or write
-     * @property baseName part of the name for the handle
-     * @property storageKey a [StorageKey]
-     * @property idGenerator used to generate unique IDs for newly stored entities.
-     */
-    suspend fun <T : Entity> createCollectionHandle(
-        mode: HandleMode,
-        baseName: String,
-        entitySpec: EntitySpec<T>,
-        storageKey: StorageKey,
-        ttl: Ttl = Ttl.Infinite,
-        idGenerator: Id.Generator = Id.Generator.newSession()
-    ): Handle =
-        idGenerator.handleName(baseName).let { name ->
-            CollectionHandle(
-                name = idGenerator.handleName(baseName),
-                entitySpec = entitySpec,
-                storageProxy = collectionStoreProxy(storageKey, entitySpec),
-                entityPreparer = EntityPreparer(name, idGenerator, entitySpec.SCHEMA, ttl, time),
-                dereferencerFactory = dereferencerFactory
-            )
-        }.let {
-            when (mode) {
-                HandleMode.Read -> when (entitySpec.SCHEMA.query) {
-                    null -> object : ReadCollectionHandle<T> by it {}
-                    else -> object : ReadQueryCollectionHandle<T, Any> by it {}
+        ttl: Ttl = Ttl.Infinite
+    ): Handle {
+        val handleName = idGenerator.newChildId(
+            idGenerator.newChildId(arcId.toArcId(), hostId),
+            spec.baseName
+        ).toString()
+        val entityPreparer = EntityPreparer<T>(
+            handleName,
+            idGenerator,
+            spec.entitySpec.SCHEMA,
+            ttl,
+            time
+        )
+        return when (spec.containerType) {
```

Comment:
this `when` is pretty long; splitting `Singleton` and `Collection` into private helper functions might be nice?

Reply:
Done